authutils/validate.go

```diff
@@ -0,0 +1,219 @@
+package authutils
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+// `fence`-specific: the `pur` field indicates the purpose for the token, which
+// may also be validated. If used, it must take one of these values.
+var ALLOWED_PURPOSES []string = []string{"id", "access", "refresh", "session", "api_key"}
+
+// JWTApplication stores the state for an application needing to validate JWTs.
+type JWTApplication struct {
+	// KeyKeys for looking up RSA public keys by ID for token validation.
+	Keys *KeysManager
+}
+
+// NewJWTApplication initializes a new application.
+func NewJWTApplication(jwkURL string) *JWTApplication {
+	keysManager := NewKeysManager(jwkURL)
+	return &JWTApplication{
+		Keys: &keysManager,
+	}
+}
+
+// Decode takes an encoded token, finds the key suitable for validating this
+// token, and returns the decoded token claims.
+//
+// NOTE that this does NOT validate the claims, only the signature.
+func (application *JWTApplication) Decode(encodedToken EncodedToken) (*Claims, error) {
+	decodedToken, err := jwt.ParseSigned(encodedToken)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the key used to sign this token.
+	//
+	// There will be multiple headers in the case of multiple signatures. This
+	// doesn't apply to our use case, so just take the key ID from the first
+	// (and what should be the only) header.
+	header := decodedToken.Headers[0]
+	kid := header.KeyID
+	key, err := application.Keys.Lookup(kid)
+	if err != nil {
+		return nil, err
+	}
+
+	claims := make(Claims)
+	if err := decodedToken.Claims(key, &claims); err != nil {
+		return nil, err
+	}
+
+	return &claims, nil
+}
+
+// checkExpiration validates the `exp` field in the claims.
+func checkExpiration(claims *Claims, now int64) error {
+	// `now` should be set with something like this:
+	//
+	//     now := time.Now().Unix()
+	tokenExp, exists := (*claims)["exp"]
+	if !exists {
+		return missingField("exp")
+	}
+	var exp int64
+	switch e := tokenExp.(type) {
+	case float32:
```

Comment:
I believe you can use commas to separate multiple case values in a single case

Reply:
When I do that it gives me this error:
```
cannot convert type e (type interface {}) to type int64: need type assertion
```

Reply:
I'm not sure what's going on here, but I wonder would go expand the cases for `tokenExp.(type)` to `tokenExp.(case)`?